feat(net):  Add nl80211 support

taschenb opened the pull request:
This replaces the deprecated Wireless Extensions with [nl80211](https://wireless.wiki.kernel.org/en/developers/Documentation/nl80211).

Fixes #277

NBonaparte commented on src/adapters/net.cpp:
```diff
@@ -324,68 +350,131 @@ namespace net {
+      // For a more intuitive result, clamp the result to a
+      // signal range, where 100% are reachable without directly
+      // sitting on top of the AP
+      const int hardware_max = -40;
```
IMO if we modify the values too much it might confuse users if they compare it to other sources.
From `nl80211_xbm_to_percent` it seems like -20 and -90 dBm are the commonly used values, so it might be best to use those instead.

patrick96 replied:
This just came up in #1834. Though [`nl80211_xbm_to_percent`](https://github.com/NetworkManager/NetworkManager/blob/master/src/platform/wifi/nm-wifi-utils-nl80211.c#L372) in the NetworkManager code says it goes from `-20` to `-90`, the way the percentage is calculated, it actually goes from `-120dBm` to `-20dBm`. `-90dBm` corresponds to 30%.

But I think the -20 to -90 range is good.

NBonaparte commented:
@taschenb Do you have any interest in continuing this PR?

patrick96 commented:
I think we should, for the time being, have both wireless_tools and nl80211 available side by side and have a config option to enable nl80211. This way we don't leave users with a broken bar, if something doesn't work.

su8 commented on src/adapters/net.cpp:
```diff
@@ -270,24 +277,43 @@ namespace net {
 
-    // Ignore interfaces in ad-hoc mode
-    if (req.u.mode == IW_MODE_ADHOC) {
+    nl_socket_modify_cb(sk, NL_CB_VALID, NL_CB_CUSTOM, scan_cb, this);
```
```c
if (0 != (nl_socket_modify_cb ...) {
handle_error
}
```

su8 commented on src/adapters/net.cpp:
```diff
@@ -324,68 +350,131 @@ namespace net {
+      if (ies_len < hdr_len || ies_len < ies[1] + hdr_len) {
+        m_essid.clear();
+      } else {
+        m_essid = string(reinterpret_cast<char*>(ies + hdr_len));
```
On top of that, ies can contain non-printable characters. So `isprint` should be used to test **ies**.

su8 commented on src/adapters/net.cpp:
```diff
@@ -324,68 +350,131 @@ namespace net {
-      m_linkquality.val = stats.qual.qual;
-      m_linkquality.max = range.max_qual.qual;
-    }
+    return status == NL80211_BSS_STATUS_ASSOCIATED || status == NL80211_BSS_STATUS_IBSS_JOINED;
```
`NL80211_BSS_STATUS_AUTHENTICATED` should be tested too.

su8 replied:
Also a **switch()** or **&** operator will be better to test against multiple cases.

taschenb commented:
I am deeply sorry for the late response.
I addressed the feedback from the code review. Thank you for that so far.

As for the coexistence with Wext, what approach would you suggest to select the interface to use? Would a simple CMake option or fallback-search for libnl in case libiw isn't found be sufficient for the time being? Or should we cleanup the network interface and expose a option in the network module?

NBonaparte commented:
A compile option should suffice. We should probably have libnl as the default and use libiw as a fallback in case the former isn't found.

taschenb commented:
In the latest update, nl80211 is now the preferred interface. In case libnl-genl-3.0 is not preset or -DWITH_LIBNL=OFF, it will fall back to Wext.

taschenb commented:
You are right, totally forgot about readding it. Is now in there as well.

NBonaparte commented:
A little nitpicky, but could you extend the other prompts to be the same length by adding more `-`s?

taschenb commented:
Good catch and not nitpicky. Everything should be and look consistent.
But the length of the prompt is still OK in general or should we try to limit this?

NBonaparte commented:
We could probably shorten it to `Include support for "internal/network" (requires libnl/libiw) -- [y/N]: ` unless there's an even shorter way.
